### ISSUE TITLE
P1032R1 Miscellaneous constexpr

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#50]
[P1032R1](https://wg21.link/P1032R1) Miscellaneous `constexpr`
~~[P1424R1](https://wg21.link/P1424R1) `__cpp_lib_constexpr`~~

WG21-P1902 "Missing Feature-Test Macros 2017-2019" further revised this, saying:
> [P1032R1] (Misc `constexpr` bits): This paper originally added a single feature test macro, `__cpp_lib_constexpr_misc`, that was since renamed by [P1424R1] to `__cpp_lib_constexpr`. Per the policy laid out in this paper, this paper proposes to remove `__cpp_lib_constexpr` and instead add separate macros for each header: `__cpp_lib_constexpr_functional`, `__cpp_lib_constexpr_iterator`, `__cpp_lib_constexpr_string_view`, `__cpp_lib_constexpr_tuple`, and `__cpp_lib_constexpr_utility` and to bump `__cpp_lib_array_constexpr`.

Feature-test macros as of WG21-N4842:
```
#define __cpp_lib_constexpr_iterator 201811L
#define __cpp_lib_constexpr_string_view 201811L
#define __cpp_lib_constexpr_tuple 201811L
#define __cpp_lib_constexpr_utility 201811L
```

The following added value is unusual (it was further increased for #51 WG21-P1065 "`constexpr` INVOKE" to `201907L`, see [SD-6][]):
`#define __cpp_lib_constexpr_functional 201811L`

The following increased value is unusual (it was originally increased for #49 WG21-P1023 "`constexpr` For `std::array` Comparisons" to `201806L`, see [SD-6][]):
`#define __cpp_lib_array_constexpr 201811L`

[SD-6]: https://isocpp.org/std/standing-documents/sd-6-sg10-feature-test-recommendations

**Comment by crackedmind** [comment]
Just for notice. I'm starting implementing this.

**Comment by crackedmind** [comment]
Decided to continue conversation from #203 here.
I started with _CONSTEXPR20 define. 

And i just realized, that i can't fully finish without compiler support. Because we need char_traits copy/move/assign specialization mark constexpr. But how i can do this without __builtin_memcpy/__builtin_memmove/__buitlin_memset ?

**Comment by StephanTLavavej** [comment]
So this is yet another feature that's blocked on #156 `is_constant_evaluated()`.

**Comment by StephanTLavavej** [comment]
`is_constant_evaluated()` is now available for Clang, and will be available for VS 2019 16.5 Preview 2 in the future, so work on this feature is unblocked (if properly guarded with `__cpp_lib_is_constant_evaluated`).
